### PR TITLE
Add NVMe-oF HII feature to timberland_1.0_final

### PR DESCRIPTION
**Pull request by trevor-cockrell**
3 commits introduced, divided by module:

- NetworkPkg
  - Addition of PcdMaxNvmeOfAttemptNumber
- NetworkPkg/Include
  - Addition of gNvmeOfConfigGuid
- NetworkPkg/NvmeOfDxe
  - Add HII/configuration forms
  - Add HII details in NvmeOfDxe README

### CODE REVIEW

**Comment by johnmeneghini** [comment]
Where are the artifacts for this MR?  I'm looking for the downloadable .zip file.

**Comment by trevor-cockrell** [comment]
> Where are the artifacts for this MR? I'm looking for the downloadable .zip file.

Had to dig -
try [artifact](https://github.com/timberland-sig/edk2/suites/13895685152/artifacts/772814781), from the [actions](https://github.com/timberland-sig/edk2/actions/runs/5390429943)

File: NetworkPkg/NvmeOfDxe/media/image048.png
Comment (johnmeneghini) [inline NetworkPkg/NvmeOfDxe/media/image048.png:1]: Can we pre-populate the `Host NQN` according to TP-TP-4126 ?

And there's not such thing at a `Host NID`. It's a `Host ID`

**Review by johnmeneghini — COMMENTED** [review]
From TP-4126 NVMe-oF Boot HostNQN and HostID:  this is requirement for the host.
```
The pre-OS boot environment and OS environment should use a fixed platform UUID to create a HostNQN
 and HostID. The implementation should use the System UUID found in the SMBIOS table. The System
 Management BIOS (SMBIOS) Reference Specification is described in DSP0134. The SMBIOS table is
 typically available to pre-OS firmware and Expansion ROM Firmware in the pre-OS boot environment as
 well as to the OS environment. 
```

**Comment by Douglas-Farley** [comment]
> From TP-4126 NVMe-oF Boot HostNQN and HostID: this is requirement for the host.

It's a `should`, not a `shall`.  I'm not opposed to following it, but it's not a requirement in that sense.

File: NetworkPkg/NvmeOfDxe/media/image048.png
Comment (Douglas-Farley) [reply NetworkPkg/NvmeOfDxe/media/image048.png:1]: > From TP-4126 NVMe-oF Boot HostNQN and HostID: this is requirement for the host.

It's a `should`, not a `shall`.  I'm not opposed to following it, but it's not a requirement in that sense.

**Comment by johnmeneghini** [comment]
> > From TP-4126 NVMe-oF Boot HostNQN and HostID: this is requirement for the host.
> 
> It's a `should`, not a `shall`. I'm not opposed to following it, but it's not a requirement in that sense.

Sorry Doug... I won't approve this change until it implements TP-4126.  It shouldn't be hard to do.

**Comment by johnmeneghini** [comment]
Bug number 2:

The current code requires the gateway address to be programmed.  This is not currently required and I always leave the gateway blank in my attempt files.  There's nothing that says you have to have a gateway programmed on your subnet, especially with a static IP.

File: NetworkPkg/NvmeOfDxe/media/image048.png
Comment (trevor-cockrell) [reply NetworkPkg/NvmeOfDxe/media/image048.png:1]: > And there's not such thing at a Host NID. It's a Host ID

Resolved in 5648f2e3826971f6b4898cfe4b71755b13ec05a1 and README updated

> Can we pre-populate the Host NQN according to TP-TP-4126 ?

Resolved in 0eb935b282b2eb6545f3ca3b72ca5b1ee46ba7af. Host ID and/or Host NQN will be pre-populated based on system UUID.

**Comment by trevor-cockrell** [comment]
> Bug number 2:
> 
> The current code requires the gateway address to be programmed. This is not currently required and I always leave the gateway blank in my attempt files. There's nothing that says you have to have a gateway programmed on your subnet, especially with a static IP.

Resolved in 2386e67d00e3418a6642c9469f1d0cbce90954bb 

Edit: New artifact [here](https://github.com/timberland-sig/edk2/suites/14777659657/artifacts/840561140)

File: NetworkPkg/NvmeOfDxe/NvmeOfMisc.c
Comment (WalonLi) [inline NetworkPkg/NvmeOfDxe/NvmeOfMisc.c]: Only one parameter in this debug call, is that coding error?

Comment (trevor-cockrell) [reply NetworkPkg/NvmeOfDxe/NvmeOfMisc.c]: Thanks! Resolved in [fa5d8f5](https://github.com/timberland-sig/edk2/pull/20/commits/fa5d8f541602b120fb5fd9c91a31097589874740)

**Comment by mwilck** [comment]
This is great! I particularly like the fact that it's still possible to pre-populate the attempt data from the EFI disk image, sparing me from having to type those UUIDs :-)

Question: I tried to enable IPv6, but it says that that's unsupported. Is that intentional?
image

If I hit ENTER at this point, it will switch back to IPv4, but the target IP address (which was an IPv4 address) will be cleared and must be re-entered.

**Comment by mwilck** [comment]
I'm wondering about the "Autoconfigure" mode. If I select that, all other configuration settings including the target subsystem ID and target IP address go away. How is this mode supposed to work?

**Comment by trevor-cockrell** [comment]
@mwilck, appreciate the feedback!!

> Question: I tried to enable IPv6, but it says that that's unsupported. Is that intentional?

No, however I've found where this is likely occurring. Fix incoming!

> If I hit ENTER at this point, it will switch back to IPv4, but the target IP address (which was an IPv4 address) will be cleared and must be re-entered.

Agreed - this is neither user-friendly nor desirable behavior. 😄 
Located issue in callback logic when switching between IP modes. Looking at reworking this to be a bit more robust/behave more intuitively.

> I'm wondering about the "Autoconfigure" mode. How is this mode supposed to work?

When `NvmeofIpMode` is set to `IP_MODE_AUTOCONFIG`, DHCP is enabled by the driver & attempted for both host and subsystem on IPv4 then IPv6 stacks. That being the case, I might have been a bit overzealous with suppressing fields in this mode.
Would it make more sense to show the (read-only) DHCP toggles in this mode as:
``` 
                                                       ^                        
   NVM Subsystem              <Disabled>                 IP4, IP6,              
                                                         Autoconfigure. In      
 > Network Device List                                   Autoconfigure mode,    
   NVMe-oF Network Device:    22:22:22:22:22:22          NVMe-oF driver will    
                                                         attempt to connect     
   Internet Protocol          <Autoconfigure>            NVMe-oF subsystem via  
                                                         IPv4 stack, if failed  
   Connection Retry Count     [3]                        then attempt IPv6      
                                                         stack.                 
   Enable DHCP                [X] // <- grayout for autoconfig
   Subsystem info from DHCP   [X] // <- grayout for autoconfig
                                                                                
   NVM Subsystem NID          7ac41e78-d9b0-4641-83c7-3d                        
                              f039ec24c6                                        
   Save Changes                                                                 
                                                       v                        
```

**Comment by mwilck** [comment]
Thanks for explaining the autoconfig mode. I recall having seen it in the old config file as well, without understanding what it was supposed to mean.

Yes, showing the greyed-out selections for DHCP will help clarifying what this is about.

Side note: The name of the feature is poorly chosen IMO, because I'll probably not be the only one who would think of IPv6 address autoconfiguration (aka Router solicitation / advertisements) in the first place, with which our "autoconfiguration" has nothing to do.
Indeed I was wondering whether IPv6 autoconfiguration without DHCPv6 was supported, which doesn't seem to be the case.

Perhaps we should use the label "both IPv4 and IPv6" instead of "autoconnect"?

**Review by johnmeneghini — DISMISSED** [review]
Red Hat is satisfied with these changes.  Thanks for working on this.

**Comment by johnmeneghini** [comment]
Internal testing of this branch has uncovered the following problems:

1. With empty OVMF_VARS.fd I found out that only the first configuration gets saved properly. When entering EFI Setup and trying to save changes for the second time, random values get lost. I was e.g. unable to save changes to the DHCP or SubsysNQN options in my boot attempt. Is this a known issue? 

2. I was having trouble saving  some other non-NBFT settings like the framebuffer resolution. Anyway,  made it working once configured fully on a first try.

3. Endianness conversion problem of the hostnqn. Consult this with the attached screenshot. 

```
# nvme nbft show --output-format=json
...
    "host":{
      "nqn":"nqn.2014-08.org.nvmexpress:uuid:12345678-abcd-abcd-abcd-
123456789012",
      "id":"78563412-cdab-cdab-abcd-123456789012",
    },
...

# cat /sys/class/nvme-fabrics/ctl/nvme0/hostnqn
nqn.2014-08.org.nvmexpress:uuid:12345678-abcd-abcd-abcd-123456789012
# cat /sys/class/nvme-fabrics/ctl/nvme0/hostid
78563412-cdab-cdab-abcd-123456789012
```

**Comment by mwilck** [comment]
> Internal testing has found more bugs which need to be addressed.

I'd still recommend to merge this now, and address any bugs or other issues separately. The addition of the HII is a huge improvement in particular for new users, even if it isn't perfect yet.

**Review by mwilck — APPROVED** [review]
I approve this PR, even if the HII code doesn't work perfectly yet.

**Comment by Douglas-Farley** [comment]
> Internal testing of this branch has uncovered the following problems:
>  [...]


John, just created a real bug for this so we can track it outside this PR in case we merge Hii and then run it to ground based on today's meeting.
see: https://github.com/timberland-sig/edk2/issues/23

**Review by johnmeneghini — APPROVED** [review]
As agreed at yesterdays meeting - Red Hat approves this PR.

**Comment by trevor-cockrell** [comment]
Closing PR in favor of: https://github.com/timberland-sig/edk2/pull/24.

**Pull request closed without merging**
